sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,317 @@
+}
+
+bool dif_plic_target_threshold_set(const dif_plic_t *plic, uint32_t priority) {
+  if (plic == NULL || !plic_priority_validate(priority)) {
+    return false;
+  }
+
+  // Write to the target specific Threshold register
+  switch (plic->target) {
+    case kDifPlicTargetIbex0:
+      reg32_write(plic->base_addr, RV_PLIC_THRESHOLD0, priority);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+bool dif_plic_irq_pending_status_get(const dif_plic_t *plic,
+                                     dif_plic_irq_src_id_t irq,
+                                     dif_plic_flag_t *status) {
+  if (plic == NULL || status == NULL) {
+    return false;
+  }
+
+  uint8_t bit_index = plic_bit_index_from_irq_id(irq);
+  ptrdiff_t offset = plic_offset_from_reg0(irq);
+  ptrdiff_t ip_offset = RV_PLIC_IP0 + offset;
+  if (reg32_get_bit(plic->base_addr, ip_offset, bit_index)) {
+    *status = kDifPlicSet;
+  } else {
+    *status = kDifPlicUnset;
+  }
+
+  return true;
+}
+
+bool dif_plic_irq_claim(dif_plic_t *plic) {
+  // If IRQ for the current |plic| state has already been claimed, but not
+  // yet processed by the caller. Every claim operation in the caller must
+  // be followed by a call to the claimed clear operation.
+  if (plic == NULL || plic_irq_claimed(plic)) {
+    return false;
+  }
+
+  // Get the IRQ ID from the target specific CC register
+  uint32_t irq_id;
+  switch (plic->target) {
+    case kDifPlicTargetIbex0:
+      irq_id = reg32_read(plic->base_addr, RV_PLIC_CC0);
+      break;
+    default:
+      return false;
+  }
+
+  // Validate the IRQ source, and determine which peripheral it belongs to
+  dif_plic_irq_src_id_t irq = (dif_plic_irq_src_id_t)irq_id;
+  if (irq == kDifPlicNoIrq) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralNone;
+  } else if (irq >= PLIC_GPIO_FIRST_IRQ_ID || irq <= PLIC_GPIO_LAST_IRQ_ID) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralGpio;
+  } else if (irq >= PLIC_UART_FIRST_IRQ_ID || irq <= PLIC_UART_LAST_IRQ_ID) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralUart;
+  } else if (irq >= PLIC_SPI_DEVICE_FIRST_IRQ_ID ||
+             irq <= PLIC_SPI_DEVICE_LAST_IRQ_ID) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralSpiDevice;
+  } else if (irq >= PLIC_FLASH_CONTROL_FIRST_IRQ_ID ||
+             irq <= PLIC_FLASH_CONTROL_LAST_IRQ_ID) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralFlashCtrl;
+  } else if (irq >= PLIC_HMAC_FIRST_IRQ_ID || irq <= PLIC_HMAC_LAST_IRQ_ID) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralHmac;
+  } else if (irq >= PLIC_ALERT_HANDLER_FIRST_IRQ ||
+             irq <= PLIC_ALERT_HANDLER_LAST_IRQ) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralAlertHandler;
+  } else if (irq >= PLIC_NMI_GEN_FIRST_IRQ || irq <= PLIC_NMI_GEN_LAST_IRQ) {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralNmiGen;
+  } else {
+    plic->claimed_irq.peripheral = kDifPlicPeripheralNone;
+    plic->claimed_irq.source = irq;
```

Comment:
`plic->claimed_irq.source = irq;` can be moved before the `if` block, since it is set in all cases. Then, the duplicate on line 293 can be removed.

Reply:
Redundant, after your suggestion